src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -454,11 +462,18 @@ void ClusterTaskManager::TasksUnblocked(const std::vector<TaskID> &ready_ids) {
 void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
                                       RayTask *task) {
   RAY_CHECK(worker != nullptr && task != nullptr);
+  if (leased_workers_.count(worker->WorkerId()) == 0) {
+    // TaskFinished has already been called on this worker.
+    return;
+  }
   *task = worker->GetAssignedTask();
+  auto sched_cls = task->GetTaskSpecification().GetSchedulingClass();
+  num_running_tasks_by_sched_cls_[sched_cls]--;
   ReleaseTaskArgs(task->GetTaskSpecification().TaskId());
   if (worker->GetAllocatedInstances() != nullptr) {
     ReleaseWorkerResources(worker);
   }
+  ScheduleAndDispatchTasks();
```

Comment:
I think we should always try to schedule and dispatch when we release args/resources since that frees up potential tasks on the dispatch queue.

Reply:
Can you revert this unless it's necessary for the PR? It would be best to keep the changes specific to the original issue.

Reply:
I mean I think this _is_ necessary since we're calling `ReturnWorker` more often now (and in the scheduling unit tests that I added there aren't many other triggers that call `ScheduleAndDispatch`).

Reply:
What I meant is that I would like to see an end-to-end example in C++ or ideally Python that shows that this is necessary.